server/s3_copy_object_action.cc

```diff
@@ -292,13 +303,76 @@ void S3CopyObjectAction::read_data_block_success() {
     return;
   }
   if (copy_failed) {
+    assert(!write_in_progress);
+
     send_response_to_s3_client();
     return;
   }
+  assert(data_blocks_read.empty());
+  data_blocks_read = motr_reader->extract_blocks_read();
+
+  size_t bytes_in_chunk_count = 0;
+
+  if (data_blocks_read.empty()) {
+    s3_log(S3_LOG_ERROR, stripped_request_id, "Motr reader returned no data");
+
+    s3_put_action_state = S3PutObjectActionState::writeFailed;
+    copy_failed = true;
+    set_s3_error("InternalError");
+
+    if (!write_in_progress) {
+      send_response_to_s3_client();
+    }
+    return;
+  }
+  // Calculating actial size of data that has just been read
+
+  for (size_t i = 0, n = data_blocks_read.size() - 1; i < n; ++i) {
+    const auto motr_block_size = data_blocks_read[i].second;
+    assert(motr_block_size == motr_unit_size);
+
+    // We can use multiplication, but something may change in the future
+    bytes_in_chunk_count += motr_block_size;
+  }
+  if (bytes_in_chunk_count >= bytes_left_to_read) {
+    s3_log(S3_LOG_ERROR, stripped_request_id, "Too many data has been read");
+
+    copy_failed = true;
+    set_s3_error("InternalError");
+
+    if (!write_in_progress) {
+      send_response_to_s3_client();
+    }
+    return;
+  }
+  bytes_left_to_read -= bytes_in_chunk_count;
+
+  auto& r_last_block_size =
+      data_blocks_read[data_blocks_read.size() - 1].second;
+
+  if (r_last_block_size >= bytes_left_to_read) {
+    r_last_block_size = bytes_left_to_read;
+    bytes_left_to_read = 0;
+  } else {
+    bytes_left_to_read -= r_last_block_size;
+  }
+  bytes_in_chunk_count += r_last_block_size;
+
+  s3_log(S3_LOG_DEBUG, request_id, "Got %zu bytes in %zu blocks",
+         bytes_in_chunk_count, data_blocks_read.size());
+
   if (!write_in_progress) {
```

Comment:
in case write_in_progress is "True", then should we log that some other write is in progress, so will write the chunk read above later. or something like this ?

This will help in debugging I guess, what do you say ?

Reply:
Agree